torchsynth/module.py

```diff
@@ -303,24 +303,38 @@ def _ramp(self, start, duration: T, inverse: bool = False) -> Signal:
         # Shape ramps.
         ramp = ramp - start_[:, None]
         ramp = torch.maximum(ramp, T(0.0))
-        ramp = ramp / duration_[:, None]
+        ramp = (ramp + EPSILON) / (duration_[:, None] + EPSILON)
         ramp = torch.minimum(ramp, T(1.0))
 
+        """
+        The following is a workaround. In inverse mode, a ramp with 0 duration 
+        (that is all 1's) becomes all 0's, which is a problem for the 
+        ultimate calculation of the ADSR signal (a * d * r => 0's). So this 
+        replaces only rows who sum to 0 (i.e., all components are zero).
+        """
+
         if inverse:
-            ramp = 1 - ramp
+            ramp = 1.0 - ramp
+            ramp[torch.sum(ramp, axis=1) == 0] = 1.0
 
         # Apply scaling factor.
         ramp = torch.pow(ramp, self.p("alpha")[:, None])
 
         return ramp.as_subclass(Signal)
 
-    def make_attack(self) -> Signal:
-        attack = self.p("attack")
+    def make_attack(self, note_on_duration) -> Signal:
+        attack = torch.minimum(self.p("attack"), note_on_duration)
         return self._ramp(torch.zeros(self.batch_size, device=attack.device), attack)
 
-    def make_decay(self) -> Signal:
+    def make_decay(self, note_on_duration) -> Signal:
+        attack = torch.minimum(self.p("attack"), note_on_duration)
```

Comment:
This is being calculated twice -- wondering if we can calculate the attack duration and decay duration outside these functions and pass the attack time and decay time into these functions instead?

Reply:
Fixed, thanks.